source/common/router/router.cc

```diff
@@ -485,7 +486,21 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   }
 
   // Fetch a connection pool for the upstream cluster.
-  Http::ConnectionPool::Instance* conn_pool = getConnPool();
+  Http::ConnectionPool::Instance* conn_pool;
+  auto url_str = headers.Host()->value().getStringView();
+  const auto parsed_authority = Http::Utility::parseAuthority(url_str.data());
+  const auto http_protocol_options = cluster_->httpProtocolOptions();
+
+  if (http_protocol_options != absl::nullopt && http_protocol_options->auto_sni() &&
+      !parsed_authority.is_ip_address_) {
+    callbacks_->streamInfo().filterState().setData(
+        Network::UpstreamServerName::key(), std::make_unique<Network::UpstreamServerName>(url_str),
+        StreamInfo::FilterState::StateType::Mutable);
+    conn_pool = getConnPool();
```

Comment:
put this outside of `if`, then you don't need `else` block

Reply:
I think this must be done after the setData() call, right? Please add a comment.

Reply:
It seems that `fromFilterState` is called on `getConnPool` so that this should refer upstream server name. so I think it is right.